### ISSUE TITLE
Clarification about the resolving of variables with language tags referenced by other variables

### ISSUE THREAD

**Issue by scr0llbaer9000** [TVTower/TVTower#1225]
So it seems with the new expressions feature the behavior of how variables with language tags referenced by other variables are resolved seems to have changed back to what it once was.

	<foo>
		<de>${bar}</de>
	</foo>
	<bar>
		<de>Hund</de>
		<en>Dog</en>
	</bar>

`${foo}` now seems to always resolve to "Hund", even when playing in English.

This behavior once had been "fixed" in https://github.com/TVTower/TVTower/pull/906 so that `${foo}` will resolve to "Dog" when playing in English, but now it seems since the introduction of the new expressions we have the old behavior again.

Even though it is still debatable which way to resolve this is more intuitive and "correct", this is a breaking change, and I'll have to re-check database entries. No problem though, have to do it anyway, as the new documentation says there is now the feature that variables can forego the language tag if the content is to be the same in any language.

So, according to the documentation, `<foo>${bar}</foo>` will now also always resolve to "Hund".

Now, for me it would actually be more intuitive if `<foo>${bar}</foo>` would always resolve to the current language (contrary to what the docu says), and with the `<de>` tag it would kinda mean that the German language is *enforced*, and the behavior can stay the same as it right now is (and contrary to what it was before the new expressions feature), but maybe that's just me. `;)`

**Comment by nittka** [comment]
I guess, the cleanest way to deal with the above example would be to create an error for English - foo is not defined. Fallbacks and best guesses are open to interpretation. Imagine the above example to contain de, en, fr and pl for bar and pl and fr for foo - which language to fall back on.

If I am right, a basic principle for the new logic is that an expression is evaluated exactly once. It is consistent that `<foo>${bar}</foo>` to the same value - the variable bar should not be language dependent.

With the old syntax, it was kind of necessary to have the fallback logic to (the same value in) one language. If you want a unique random city, you cannot have 
%STATIONMAP:RANDOMCITY% in each language tag but because there was no global variable definition, using a single language was the workaround.

To be honest, rather than introducing a shorthand logic, I'd require the database to either copy definitions for the required languages or use a global value in case the result is exactly the same for all languages.
Note that the naive validation "each variable must be global or defined for all languages" does not work. It is well possible that for one language grammar specific variables are required that are not needed for other languages.

**Comment by nittka** [comment]
(Note that a side effect of this change is that saved games will be "broken" for converted entries that relied on that fallback logic for certain languages.)

**Comment by scr0llbaer9000** [comment]
About the randomcity it could *theoretically* and *eventually* also be that the generator returns a city that different names in various languages (like Munich vs München). There currently would be no way to use this. Without language tags the returned city name would be "static", with one language tag it would be in the language of the specified city, with several language tags it would return different cities for each language.

**Comment by nittka** [comment]
@GWRon : `<foo><de>${bar}</de></foo>` and `<foo>${bar}</foo>` are effectively the same at the moment. Both lead to the same value for all languages.
It could be argued, that the first version should be a shorthand for "use the same definition in all languages" (rather than use the same value in all languages).
From briefly looking at the code, there is no way to store a localized string as global (without a specifix language, -1 is interpreted as current language).

The fix in the PR mentioned was in the template variables resolution. For each language a copy of the original value was created. This code has been completely replaced by the script expression resolution.

If we do not want to change the framework code (localized string, template variable and its resolution etc.) but still allow the above distinction (which would make sense from a database point of view - readability, no meaningless duplicates), an option would be to automatically create the copies when reading the database:
* if no language tag is present, create a single localized string for the default language (yielding the same result for all languages)
* if a single language tag is present, store a copy of that string for all languages
* proceed just like now

**Comment by GWRon** [comment]
> `<foo><de>${bar}</de></foo>` and `<foo>${bar}</foo>` are effectively the same at the moment. Both lead to the same value for all languages.

Yes and I thought it was desired to behave this way. If nothing "for my language" was defined differently, use what is there. If there was `<foo><de>${bar}</de><en>ENGLISH</en><it>${bar}</it></foo>` and you request "foo" for the Polish language, it should return "ENGLISH", if you asked for German text, then the expression would need to be evaluated. The Italian text would be the same as the German, as they both use the same "variable". But the variable can be a localized string too - in that case it could contain Italian and/or German variants, while an "expression" inside of the variable would only be evaluated once (eg a stationmap-randomcity call)


> if a single language tag is present, store a copy of that string for all languages

You mean it should copy the un-evaluated `${bar}` in each language (and what is "all languages" - how to know which one to use?)?

**Comment by nittka** [comment]
> Yes and I thought it was desired to behave this way. If nothing "for my language" was defined differently, use what is there.

But in scr0llbaer's example bar is indeed a "localized" string containing an English and a German variant, so the English foo should be resolved to "dog" with your interpretation. This would indicate a bug in the template variable/expression resolution - variables should not return a simple text token but a localized string token!?

If it is not a bug, copying `${bar}` for each language in the the localized string should work as expected. It is always the same variable but for each language it is evaluated separately (with the existing fallback).
If this did not work already most scripts would be broken (`<title><de>${title}</de><en>${title}</en>...</title>`).

**Comment by GWRon** [comment]
You are right - I did a misassumption.

> If there was <foo><de>${bar}</de><en>ENGLISH</en><it>${bar}</it></foo> and you request "foo" for the Polish language, it should return "ENGLISH", if you asked for German text, then the expression would need to be evaluated. The Italian text would be the same as the German, as they both use the same "variable". But the variable can be a localized string too - in that case it could contain Italian and/or German variants,


OK, so I tried this now:
```xml
		<scripttemplate product="1" licence_type="1" guid="test-scripttemplate-1">
			<title>
				<de>German title</de>
				<en>English title</en>
			</title>
			<description>
				<de>${foo} German description.</de>
				<en>${foo} English description.</en>
				<es>${foo} Spanish description.</es>
				<fr>${foo} French description.</fr>
			</description>

			<variables>
				<bar>
					<de>GERMAN</de>
					<fr>FRENCH</fr>
					<en>ENGLISH</en>
				</bar>
				<foo>
					<de>${bar} (DE)</de>
					<en>ENGLISH (EN)</en>
					<fr>${bar} (FR)</fr>
					<es>${bar} (ES)</es>
				</foo>
			</variables>

			<jobs>
				<job index="0" function="1" required="1" />
			</jobs>

			<data flags="1" production_broadcast_flags="512" target_groups="7" />
			<genres maingenre="104" />
			<blocks min="2" max="3" slope="40" />
			<price min="10000" max="18000" slope="40" />
			<potential min="30" max="55" slope="45" />
			<outcome min="20" max="35" slope="40" />
			<review min="25" max="45" slope="40" />
			<speed min="45" max="60" slope="50" />
		</scripttemplate>
```

Output via:
```
	For local lang:String = eachin ["de", "en", "fr", "es"]
		TLocalization.SetCurrentLanguage(lang)
		print "lang="+lang+": T=" +script.GetTitle()
		print "lang="+lang+": D=" +script.GetDescription()
	Next
```

```
lang=de: T=German title
lang=de: D=GERMAN (DE) German description.
lang=en: T=English title
lang=en: D=ENGLISH (EN) English description.
lang=fr: T=English title
lang=fr: D=FRENCH (FR) French description.
lang=es: T=English title
lang=es: D=ENGLISH (ES) Spanish description.
```
So title correctly uses "English" (as previously defined fallback language). French uses correct resolved "FRENCH"-bar and the Spanish one too.

Are you talking about something different?

**Comment by GWRon** [comment]
Think i finally understood the issue @scr0llbaer9000 explained.

So instead of:
```XML
			<variables>
				<bar>
					<de>GERMAN-bar</de>
					<fr>FRENCH-bar</fr>
					<en>ENGLISH-bar</en>
				</bar>
				<foo>
					<de>${bar} (DE)</de>
					<en>ENGLISH (EN)</en>
					<fr>${bar} (FR)</fr>
					<es>${bar} (ES)</es>
				</foo>
			</variables>
```
the entry would look like:
```XML
			<variables>
				<bar>
					<de>GERMAN-bar</de>
					<fr>FRENCH-bar</fr>
					<en>ENGLISH-bar</en>
				</bar>
				<foo>
					<de>${bar} (DE)</de>
				</foo>
			</variables>
```
and the old code would return for `foo`:
de: `GERMAN-bar (DE)`
en: `ENGLISH-bar (DE)`
fr: `FRENCH-bar (DE)`
right?

If so, then I would consider this "not logical" because "foo" is (in the example) explicitely defined as a variable with only one "localization" ("de").

If we now fixed it by checking if the variables used in the "text" (`${bar} (DE)` contains the variable `bar`... so this variable would need lookup) and create variants for any locale-key used (so here: de, en ,fr) this would become more complicated as soon as multiple variables are used - and each of them with varying locale keys.
As nittka already pointed out the script expression evaluation only cares for a variable once, it does not look up for further variables down the line (on the same "level"). So `<de>${bar} (DE) ${deadbeef}</de>` would care for `bar` first and somewhen later for `deadbeef`, during evaluation of `bar` there is no information about `deadbeef` available yet.


> If it is not a bug, copying ${bar} for each language in the the localized string should work as expected.

As said - how to know which language/locale keys to support on copying?

**Comment by GWRon** [comment]
@nittka 
For now eg the scripts replace their description/title-content this way:
```
	Method _ReplaceScriptExpressions:TLocalizedString(text:TLocalizedString, useTime:Long = 0)
		Local result:TLocalizedString = text.copy()

		For Local langID:Int = EachIn text.GetLanguageIDs()
			Local valueOld:String = text.Get(langID)
			Local context:SScriptExpressionContext = new SScriptExpressionContext(self, langID, Null)
			Local valueNew:TStringBuilder = GameScriptExpression.ParseLocalizedText(valueOld, context)
			If valueOLD <> valueNew.Hash()
				result.Set(valueNew.ToString(), langID)
			EndIf
		Next

		Return result
	End Method
```

This means if a title was only defined for `<de>` (because of only using variables in there and the hope for multi-language support in the other variables), then script expression will only be evaluated for the "de" locale "path".

So we need to find a way in advance, which languages the "databases" support ... and thus make this a "minimum" of what above function should loop over (so a mix of localizedstring-provided locales, and the "minimum").

As the locale is passed via the context to the expression handler, it should pass it around too when retrieving locales from the various variables.

**Comment by nittka** [comment]
I think, a test case to check whether there is a bug is indeed
```
<foo>
	<de>${bar}</de>
</foo>
<bar>
	<de>Hund|Katze</de>
	<en>Dog|Cat</en>
</bar>
```
where foo and bar are used in titles and descriptions. Start with English to generate the titles and then switch to German. If Hund/Katze do not appear, it is a bug! The database entries clearly state that German foo and bar are defined.
If the entry `<en>${bar}</en>` existed everything should be fine ( - but somewhat contrary to the claim ${bar} ist not evaluated only once but once for each language).

> As said - how to know which language/locale keys to support on copying?

I repeat my proposal from above (done during loading the database, not in the expressions themselves):
* If the database entry does not contain a language tag, use a single entry with the default language. It will behave as now - an identical result for each language independent of any variables that might have localizations (it is the database maintainers decision to use global or localized variables).
* If the database entry contains a single language tag, copy that content for each supported language (or extend the localized String with a "same base value for each language" variable). That way, even without changing the evaluation logic, the game should behave as before. For each language a separate evaluation is attempted.
* If the database contains some but not all language tags - little the game can do, fix the database.

Note that it will not be possible to use a global variable as shortcut for "use the same base for all languages" - the unique randomcity use case could not be achieved. How about
```
<foo>
	<all>${bar}</all>
</foo>
```
to make clear that the same base should be used for all languages.

**Comment by GWRon** [comment]
> If the database entry contains a single language tag, copy that content for each supported language (or extend the localized String with a "same base value for each language" variable)

Again: how to know what languages are "supported" - do you really want to create a string for ANY language we have as "interface language" ?
I normally would have thought it requires a 2 step parsing of the database: first is to scan for information (what language tags are used) and in the second the actual loading. This would be required as during loading the elements are already creating eg. the programme licences.

**Comment by GWRon** [comment]
I need to append another thing:
As said we need to know which language/locale keys to iterate over.

Variables can be inherited from parent in other files. So eg someone extends a script to contain more elements and only adds texts in English and German while the Original contained only English, Polish and French.
Now that newly added scripts refer to variables defined in the parent script of the other database file.

Which languages do we need to support then? 
- English, Polish and French
- or: English, Polish, French and German
- or: English and German

And remember you need to find a way to see what locales are to use (so the custom appended script needs to identify that the original parent script was offering English, Polish and French).

---- 
Normally I would have suggested to simply only support "fallback language" (English) plus "Client language" (so German, French, Polish ...). If you have a savegame of a foreign player you could simply continue playing in "English" and should be able to follow whatever is there.
If someone wants to play in Polish and there is no translation they will be receiving English anyways, no need to include German, Italian, French ... then anymore.


So during loading stuff in you would evaluate expressions just for the fallback language and the display/client language. And yes, this means if you accidentally start your new game in "English", then this will be the only language we need to store and you cannot change the language for programmes, news, ... anymore (except you force reloading of the databases with... but this only works for non-dynamic-stuff then)

**Comment by GWRon** [comment]
The bug test case works as expected if you iterate over EN and DE in the replaceScriptExpression methods in scripts/news classes... There it by defaults iterates over the language locales defined in the localizedString (which in our case is DE only).

This is why you need to know the languages "at all" (independent from the currently processed TLocalizedString).

And this means if someone translated everything on their own, they might not have touched single language entries just containing a variable reference ... And there we would not know that their custom language exists and would still only translate "de" ...or maybe en,de,pl.

**Comment by GWRon** [comment]
Quick mockup...
![image](https://github.com/user-attachments/assets/65df8411-b349-4176-b4d0-ad0e4559d200)
So the left is eg the "title:TLocalizedString" which is requested to get "evaluated".
That `<title>` is defined for two languages: DE and EN.

The current algorithm (eg. game.production.script.bmx - _ReplaceScriptExpressions) iterates over every of these locale keys there and evaluates the expression in the given context of "this language" - so once for "DE" and once for "EN".

During evaluation it has to evaluate VAR1 ... which happens to only contain "DE" values. But when asked to "get the content" with "DE" it will return its value - and the same value it returns for "EN". If that variable contains another variable/expression that one is evaluated too but the resolved value then is stored for the given language. So after evaluation of "DE" it will most probably already contain the resolved one (need to check that) and return the already evaluated one - or it would start again with the "unresolved" text (which is still the "de" one, but might still contain the expressions there).
VAR2 contains DE and EN again - and thus both paths would be "used".
VAR3 contains "FR" but this one is not evaluated at all as the "initial" paths are defined by the text (so only DE and EN are checked).

To enable the FR-path we would need to know in advance that FR is even an option - and this is what I was asking all the time.

**Comment by nittka** [comment]
I did not understand the first time you asked, but I did a while back. And maybe, I did not make my proposal clear enough. I try again. In your quick mockup, I simply don't care if someone wrote a database entry that contained a new language in the last level. If the language is not present in the entry point, there is no guarantee for the outcome in the game. If someone used a global variable (no language tag) containing a variable that might lead to different languages, there is no guarantee for the outcome (global is intended to be identical for all languages). Please write consistent database entries!

What I care (a little) about is preventing repetition in the database xml file (<title><de>${title}</de><en>${title}<en>...</title>). So the suggestion is when processing the xml for a variable`<var><all>content</all></var>` to store content for the fixed set of languages currently supported by the official database - en, de, pl. (I don't care that there are 6 movies with French titles and 3 with Italian ones, movies do not contain variables, only templates do.)

Using `all` instead of the "any single language tag" heuristic would have advantages, I don't want to elaborate on...
If someone now translated the entire official database to Norwegian - great, he'll probably want to play in Norwegian and won't care if something broke in German. In the next release the set of official languages could easily be changed.

**Comment by GWRon** [comment]
There is no need to have a set of languages ..if all entries for title, description... would contain all language keys we care of.

We only need that predefined set for all entries trying to do a shortcut.

```
<title>
 <de>Titel</de>
</title>
```
Would be then just be localized in German.

But what we simply could do is to have a fixed set defined (somewhere...?) like for now "de,en,pl". And this is the minimum of locale keys it evaluates. So if one extends the title to contain "swe", then it would also try to localize into Swedish.

It is always the entry point who decides what at least to localize ..and we would now just merge that with a fixed set. 
At the end this defines the routes we use for evaluation ..but there we then evaluate from "inner to outer" (in the image.. right to left).

So ..where to define the minimum language set?

**Comment by nittka** [comment]
> So ..where to define the minimum language set?

I thought this is obvious. database.bmx does transforms from xml to model. This is where the language set is applied (and defined). For the moment it is overkill to make it configurable.

I don't think the framework itself is the place to fix this. Simply duplicate the entries.

**Comment by GWRon** [comment]
So you want (and have described it roughly already) to extend all loaded in TLocalizedString within game.database.bmx to contain at least all our minimum locale keys (copy from fallback if needed). This to avoid changes in other parts of the framework (eg script replace expression functions...).
Right?

If so, we just need to clarify if we just extend title, description,... Or if we also extend all variables (so all TLocalizedString). The latter of course increases savegames/memory usage more than the other (dunno how much, but it increases). But it avoids having to think about "have i missed a variable".

**Comment by nittka** [comment]
Review the database
* `<var>value</var>` for for entries identical for all languages
* `<var><all>${otherVar}</all></var>` for entries whose definition would be identical for all languages

in database.xml expand language tag `all` to the currently supported languages. If `<var><pl>...` is defined after all, it might even work to override single language entries. Otherwise all language tags would have to be defined even if one is slightly different.

**Comment by nittka** [comment]
Sprachtag `<all>...</all>` wird jetzt unterstützt. Eine erste Aufgabe wird nun sein, die Datenbank zu durchkämmen und alle Einzelspracheinträge darauf zu untersuchen, ob all oder "global" gemeint ist.

**Comment by nittka** [comment]
@scr0llbaer9000 : with a local validation implementation I can conveniently check where only one language is defined or where all languages use the same pattern. Within a couple of days, I could transform these places to either a global definition or use "all" as single language and prepare a PR for you to review. Would you be interested?

**Comment by scr0llbaer9000** [comment]
ok I guess it's better when you make the changes in an automated way first.
